File: frontend/src/types/attendance.ts
export interface AttendanceRecord {
  id: string;
  tenant_id: string;
  driver_id: string;
  shift_id: string | null;
  date: string;
  status: string;
  scheduled_start: string | null;
  actual_start: string | null;
  late_minutes: number;
  source: string;
  selfie_url: string | null;
  location_lat: number | null;
  location_lng: number | null;
  notes: string | null;
  created_at: string;
}

export interface AttendanceCreate {
  driver_id: string;
  shift_id?: string;
  date: string;
  status: string;
  scheduled_start?: string;
  actual_start?: string;
  late_minutes?: number;
  source?: string;
  notes?: string;
}

export interface AttendanceUpdate {
  status?: string;
  actual_start?: string;
  late_minutes?: number;
  notes?: string;
}

export interface AttendanceFilters {
  date_from?: string;
  date_to?: string;
  driver_id?: string;
  status?: string;
  search?: string;
  page?: number;
  per_page?: number;
}

export interface AttendanceSummary {
  date: string;
  summary: Record<string, number>;
  attendance_rate: number;
  avg_late_minutes: number;
}

File: backend/app/schemas/attendance.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel


class AttendanceCreate(BaseModel):
    driver_id: uuid.UUID
    shift_id: uuid.UUID | None = None
    date: date
    status: str
    scheduled_start: datetime | None = None
    actual_start: datetime | None = None
    late_minutes: int = 0
    source: str = "manual"
    notes: str | None = None


class AttendanceUpdate(BaseModel):
    status: str | None = None
    actual_start: datetime | None = None
    late_minutes: int | None = None
    notes: str | None = None


class AttendanceResponse(BaseModel):
    id: uuid.UUID
    tenant_id: uuid.UUID
    driver_id: uuid.UUID
    shift_id: uuid.UUID | None
    date: date
    status: str
    scheduled_start: datetime | None
    actual_start: datetime | None
    late_minutes: int
    source: str
    selfie_url: str | None
    location_lat: float | None
    location_lng: float | None
    notes: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class AttendanceSummaryResponse(BaseModel):
    date: str
    summary: dict[str, int]
    attendance_rate: float
    avg_late_minutes: float
